### DISCUSSION TITLE
PlotWidget and PlotItem instance not plotting data in multiple inheritance

### DISCUSSION THREAD

**Discussion by yudevan** [pyqtgraph/pyqtgraph#2908] [Q&A]
Hi,
I have following code

<pre class="notranslate">
class CurveItem(PlotItem):
    index = 0
    instances = []
    RegionUpdate  = QtCore.pyqtSignal(int, int)
    def __init__(self,*args, **kwargs ):
        super(CurveItem, self).__init__(*args, **kwargs )
        self.instances.append(self)
        self.showGrid(x=True, y=True)
        self.legend =LegendItem()
        self.legend.setParentItem(self)
        self.legend.setOffset((150,50))
        self.addLegend()
        self.SeaRegion = LinearRegionItem(brush=mkBrush((0,100,255,15)))
        self.SeaRegion.setZValue(100)
        self.SeaRegion.setRegion([int(datetime.timestamp(datetime.now()) - 21600), int(datetime.timestamp(datetime.now()) + 21600)])
        #self.SeaRegion.sigRegionChanged.connect(self.regionMoved)
        self.addItem(self.SeaRegion, ignoreBounds=True)
        self.todaymark = InfiniteLine(angle=90, movable=False, pos=(datetime.timestamp(datetime.now())), name="Now", pen="r")
        self.todaymark.setFocus()
        self.addItem(self.todaymark, ignoreBounds=True)
        #self.viewbox = self.getViewBox()
        self.setAutoVisible(y=True)
</pre>

and I have QtWidget baseclass module with following code in the __init__ section:
<pre class="notranslate">

        self.rawCurveAxis = DateAxisItem()
        self.rawMagCurve = CurveItem(name="Raw Plot", axisItems={'bottom':self.rawCurveAxis})
        self.rawMagCurve.getAxis("left").setLabel('Current Magnitude', color="#2E2EFE", units='knots')
        self.rawMagCurve.setLabel(DockablePlot.Position.BOTTOM, "Time")   
        self.rawPlotWidget = PlotWidget(axisItems = {
            'left': AxisItem(orientation='left', units="knots"),
            'bottom': DateAxisItem(),
            'right': AxisItem(orientation='right',pen='#FE2E64',units="{}".format(chr(176)))
            }
        )
        self.rawPlotWidget.showGrid(x=True, y=True)
        self.plotArea.rawPCPanel.addWidget(self.rawPlotWidget)
</pre>

and I defined 2 methods to perform plotting using defined instan
<pre class="notranslate">
def PlotMagPW(self,  x_y_data,name:str="default", *args, **kwargs):
        if name == "default":
            name = f"{self.metadata.name} Raw"
        else:
            name = f"{name} Magnitude"
        
        if name in self.rawplot.keys():
            self.rawplot[name].setData(x_y_data)
        else: 
            self.rawplot[name]=self.rawPlotWidget.plot(x_y_data)
        return self.rawplot[name]

    def PlotDirPW(self,  x_y_data,name:str="default", *args, **kwargs):
        if name == "default":
            name = f"{self.metadata.name} Raw"
        else:
            name = f"{name} Magnitude"
        if name in self.rawplot.keys():
            self.rawplot[name].setData(x_y_data)
        else: 
            self.rawplot[name]=self.rawPlotWidget.plot(x_y_data)
        return self.rawplot[name]

   def PlotRawMag(self,  x_y_data,name:str="default", *args, **kwargs):
        if name == "default":
            name = f"{self.metadata.name} Raw"
        else:
            name = f"{name} Raw Magnitude"
        if name in self.rawplot.keys():
            self.rawplot[name].setData(x_y_data)
        else: 
            self.rawplot[name]=self.rawMagCurve.plot(x_y_data, name=name, *args, **kwargs)
            self.rawMagViewbox.addItem(self.rawplot[name])
        return self.rawplot[name]
    
    def PlotRawDir(self,  x_y_data,name:str="default", *args, **kwargs):
        if name == "default":
            name = f"{self.metadata.name} Raw"
        else:
            name = f"{name} Raw Direction"
        if name in self.rawplot.keys():
            self.rawplot[name].setData(x_y_data)
        else:
            self.rawplot[name]=self.rawMagCurve.plot(x_y_data, name=name, *args, **kwargs)
            self.rawDirViewbox.addItem(self.rawplot[name])
            
        return self.rawplot[name]
</pre>

and then I created subclass from that baseclass and tried to perform plotting with following code:
<pre class="notranslate">
def plotRawdata(self, widget):
        currentDf=self.reader.loadWithSpecs(self.metadata.params) # Loading CSV file via pandas
        currentDf.dropna()
        speedPen = self.colormap.getNextPen(width=4)
        directionPen = self.colormap.getNextPen(width=4)
        epoch = self.gpstime2numpy(currentDf.iloc[:,0].values) # Convert gps epoch to numpy epoch
        datetime = currentDf.iloc[:,1] # New dataframe containing column number 1
        mag = currentDf.iloc[:,2] # Create new dataframe containing magnitude column number 2 
        dirn = currentDf.iloc[:,3] # Create new dataframe containing direction number 0 and 3
        print(f"{self}\n",currentDf.iloc[1:,0].describe())
        print(f"{self}\n",mag.describe())
        magdata = np.stack([ epoch, mag.to_numpy() ], axis=-1)
        dirdata = np.stack([ epoch, dirn.to_numpy() ], axis=-1)
        self.PlotRawMag(magdata, pen=speedPen, name=f"{self.reader.name}") # Plotting using PlotItem instance
        self.PlotRawDir(dirdata, pen=directionPen,name=f"{self.reader.name}") # Plotting using PlotItem instance
        self.PlotMagPW(magdata, pen=speedPen, name=f"PW {self.reader.name}") # Plotting using PlotWidget instance
        self.PlotDirPW(dirdata, pen=directionPen,name=f"PW {self.reader.name}") # Plotting using PlotWidget instance

        self.rawDirViewbox.setGeometry(self.rawMagViewbox.sceneBoundingRect())
        self.rawMagCurve.setRegion(
                [int(epoch[-1] - 7200), int(epoch[-1] + 7200)]
        )
        # Raise and set display 
        self.plotArea.rawPanel.raiseDock()

</pre>

When I use the code above there are no plot being produced, neither an error which make it so difficult to debug. I have been trying to run it through ipdb but found nothing.
but when I put these code into  plotRawdata:
<pre class="notranslate">
       # below method is working but why above codes not working
        if not self.pw:
             self.pw = pg.PlotWidget(axisItems = {'bottom': pg.DateAxisItem()})
             self.plotArea.vlayout.addWidget(self.pw)
        
         self.pw.plot(magdata,pen=speedPen, name=f"PW {self.reader.name}")
         self.pw.plot(dirdata,pen=directionPen,name=f"PW {self.reader.name}")
</pre>

The plot produced successfully without problem. I have same implementation method using PlotItem and it showing same behaviour. 
 I was currious if it because I produced the data from pandas Series using to_numpy() method.  because I have same data structure produced by pure numpy and fed it into plotRawdata I am able to make plot with no problem. 

I did comparation by dumping the data produced by both pandas and numpy, but I didn't see any differences.

could please anyone tell me what I am missing.

thank you

**Comment by pijyoi** [comment]
You did not supply your `PlotItem`-derived `CurveItem` as an argument to `PlotWidget` when instantiating the latter.
https://pyqtgraph.readthedocs.io/en/latest/api_reference/widgets/plotwidget.html

But why derive from `PlotItem` at all?

**Comment by yudevan** [comment]
Hi Pijyoi,

thanks for replying 

> You did not supply your `PlotItem`-derived `CurveItem` as an argument to `PlotWidget` when instantiating the latter. https://pyqtgraph.readthedocs.io/en/latest/api_reference/widgets/plotwidget.html

I deliberately did not provide PlotItem to the PlotWidget as it will create one by default when we instantiate it, but yes I did try that as well, and it did behave the same, when I supply ndarray created directly from numpy it is plotting with no problem. but when the ndarray created from pandas Dataframe using to_numpy() method it didn't plot and didn't give error either, this what confuse me. I have tried to debug it but got no luck.
> 
> But why derive from `PlotItem` at all?

The idea of subclassing PlotItem is that so I can have all the components I need such as Region, Legends, Vertical mark etc on every instance without creating it over and over. the original idea was only using PlotItem, the PlotWidget is just an attempt for troubleshooting why it wasn't plotting. and it seems it behave the same between both of them. 
it will successfully plotting data converted from pandas if I create the PlotWidget instance and call plot method from the same function like this :

<pre>
def PlotRawData(self,widget):
        if not self.pw:
             self.pw = pg.PlotWidget(axisItems = {'bottom': pg.DateAxisItem()})
             self.plotArea.vlayout.addWidget(self.pw)
        
         self.pw.plot(magdata,pen=speedPen, name=f"PW {self.reader.name}")
         self.pw.plot(dirdata,pen=directionPen,name=f"PW {self.reader.name}")
</pre>

Thanks

**Reply by pijyoi** [reply]
Could you provide an MWE that demonstrates the issue that you are facing?

**Comment by yudevan** [comment]
Hi pijyoi, 

I have updated the code little bit, it is now adding the legend but still not the plot here are the code
<pre>
def plotRawdata(self, widget):
        currentDf=self.reader.loadWithSpecs(self.metadata.params)
        currentDf.dropna()
        speedPen = self.colormap.getNextPen(width=4)
        directionPen = self.colormap.getNextPen(width=4)
        epoch = self.gpstime2numpy(currentDf.iloc[:,0].to_numpy(copy=True)) # Convert gps epoch to numpy epoch
        datetime = currentDf.iloc[:,1].to_numpy(copy=True) # New dataframe containing column number 1
        mag = currentDf.iloc[:,2].to_numpy(copy=True) # Create new dataframe containing magnitude column number 2 
        dirn = currentDf.iloc[:,3].to_numpy(copy=True) # Create new dataframe containing direction number 0 and 3
        magdata = np.stack([ epoch, mag ], axis=-1)
        dirdata = np.stack([ epoch, dirn ], axis=-1)
        
        # if not self.rawMagCurve:
        #     self.rawMagCurve=RawCurrentPlotItem(name="Raw Plot", axisItems={'bottom':self.rawCurveAxis})
        self.PlotRawMag(magdata, pen=speedPen, name=f"{self.reader.name}")
        self.PlotRawDir(dirdata, pen=directionPen,name=f"{self.reader.name}")
        
        #self.rawDirViewbox.setGeometry(self.rawMagViewbox.sceneBoundingRect())
        self.rawMagCurve.setRegion(
                [int(epoch[-1] - 7200), int(epoch[-1] + 7200)]
        )
        # Raise and set display
        self.loginfo(self.rawplot)
        self.plotArea.rawPanel.raiseDock()
</pre>

and here is the output from those code

![Screenshot from 2023-12-29 08-01-55](https://github.com/pyqtgraph/pyqtgraph/assets/4899328/9b08ee07-9b27-4964-ad7f-a134bb80e022)
The data is in the format of pandas dataframe converted into numpy array using to_numpy method.

and here is the example of the plot generated from pure numpy. these data plotted at the same PlotItem window as used above 
![Screenshot from 2023-12-29 08-03-02](https://github.com/pyqtgraph/pyqtgraph/assets/4899328/b2c5c296-a8a5-4883-b163-bd692ad9a44a)

Sinusoid plot generated from pure numpy.

here is the code generating the plot above
<pre>

        plotFeather = np.stack([datecolumn.tolist(), feathercolumn.tolist()], axis=-1)
        print("1_overlaysine.OverlaySine.plot ", type(plotFeather), plotFeather.dtype, plotFeather.shape, np.info(plotFeather))
        self.PlotFeather(plotFeather, pen=pencolor, name="Sinusoid")
        self.PlotRawDir(plotFeather, pen=self.colormap.getNextPen(width=3), name="Sinusoid")
        self.PlotRawMag(plotFeather, pen=self.colormap.getNextPen(width=3), name="Sinusoid")
</pre>
I am so confused what is causing this problem, as the array generated from both code are valid numpy array, and pyqtgraph didn't complain anything.

thank you
